src/wutil/gettext.rs

```diff
@@ -120,25 +117,105 @@ fn wgettext_impl(text: MaybeStatic) -> &'static wstr {
-/// this is checked in debug mode.
-pub fn wgettext_static_str(s: &'static wstr) -> &'static wstr {
-    wgettext_impl(MaybeStatic::Static(s))
+/// A string which can be translated.
+/// The wrapped string itself is the original, untranslated version.
+/// Use [`TranslatableString::translate`] to obtain the translation.
+///
+/// Do not construct this type directly.
+/// For string literals defined in fish's Rust sources,
+/// use the macros defined in this file.
+/// For strings defined elsewhere, use [`TranslatableString::from_external_source`].
+/// Use this function with caution. If the string is not extracted into the gettext PO files from
+/// which fish obtains translations, translations will not work.
+#[derive(Debug, Clone)]
+pub enum TranslatableString {
+    Static(&'static wstr),
+    Owned(WString),
+}
+
+impl TranslatableString {
+    /// Create a [`TranslatableString`] from a string which is not from fish's own Rust sources.
+    /// Translation will only work if this string is extracted into the translation files some
+    /// other way.
+    pub fn from_external_source(s: WString) -> Self {
+        Self::Owned(s)
+    }
+
+    /// Get the translation of a [`TranslatableString`].
+    /// If original string is empty, an empty `wstr` is returned,
+    /// instead of the gettext metadata.
+    pub fn translate(&self) -> &'static wstr {
+        match self {
+            Self::Static(s) => {
+                if s.is_empty() {
+                    L!("")
+                } else {
+                    wgettext_impl(MaybeStatic::Static(s))
+                }
+            }
+            Self::Owned(s) => {
+                if s.is_empty() {
+                    L!("")
+                } else {
+                    wgettext_impl(MaybeStatic::Local(s))
+                }
+            }
+        }
+    }
+}
+
+impl std::fmt::Display for TranslatableString {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.translate())
+    }
+}
+
+/// This macro takes a string literal and produces a [`TranslatableString`].
+/// The essential part is the invocation of the proc macro,
+/// which ensures that the string gets extracted for translation.
+#[macro_export]
+macro_rules! translatable_str {
+    ($string:literal) => {
+        $crate::wutil::gettext::TranslatableString::Static(widestring::utf32str!(
+            fish_gettext_extraction::gettext_extract!($string)
+        ))
+    };
 }
+pub use translatable_str;
 
-/// Get a (possibly translated) string from a non-literal.
-/// This truncates at the first NUL character.
-pub fn wgettext_str(s: &wstr) -> &'static wstr {
-    wgettext_impl(MaybeStatic::Local(truncate_at_nul(s)))
+/// Macro for declaring string consts which should be localized.
+#[macro_export]
+macro_rules! translatable_consts {
+    (
+        $(
+            $(#[$attr:meta])*
+            $vis:vis
+            $name:ident
+            $string:literal
+        )*
```

Comment:
I didn't expect the repetition here but I guess I don't see any problems with this.
LSP go-to-definition still works fine.